lib/friends/introvert.rb

```diff
@@ -777,5 +779,52 @@ def thing_with_name_in(type, text)
     def bad_line(expected, line_num)
       raise FriendsError, "Expected \"#{expected}\" on line #{line_num}"
     end
+
+    def decide_default_location_output(activity)
+      @activities = stable_sort(@activities)
+      existing_activities = @activities - [activity]
+
+      str = "Default location"
+      if activity != @activities.first
+        str += " from #{earliest_consecutive_default_location_activity_date(activity, existing_activities)} to"
+        str += " #{next_activity_date_with_different_default_location(activity, existing_activities) || 'present'}"
+      end
+
+      str += " already" if last_default_location_same_as_added_default_location?(activity, existing_activities)
+      str += " set to: \"#{activity.default_location}\""
+
+      @output << str
+    end
+
+    def last_default_location_same_as_added_default_location?(activity, existing_activities)
+      last_default_location_activity = existing_activities.select { |a| a.default_location && (a.date < activity.date) }.first
+      if last_default_location_activity
+        last_default_location_activity.default_location == activity.default_location
+      end
+    end
+
+    def earliest_consecutive_default_location_activity_date(activity, existing_activities)
+      # binding.pry
+      earlier_default_location_activities = existing_activities.select { |a| a.default_location && (a.date < activity.date) }
```

Comment:
I don't think you need the parentheses here, right? Also, using a `partition` as in my version helps to avoid having to do these date comparisons in multiple places. Thoughts?

Reply:
Parentheses can definitely be removed, and yes using the `partition` method is much more efficient 👍